### ISSUE TITLE
Update Lenovo 40B0 Thunderbolt 4 Dock firmware

### ISSUE THREAD

**Issue by VeldoraTheDragon** [fwupd/firmware-lenovo#335]
We use the Lenovo Thunderbolt 4 docks for our production environment.
Unfortunately, the dock's firmware is outdated on lvfs: 10.13.
The latest version is 1.0.15.

We would like to ask you to update the firmware to the latest available version so that it can be updated via fwupdmgr, so that no additional windows-vm is required to update the firmware.

We would appreciate it if the firmware could be released more frequently in the future. (e.g one week after official lenovo release)

Thanks in advance.

**Comment by mrhpearson** [comment]
Acked and forwarded to the dock team.
They are working on the update.

**Comment by ralfbergs** [comment]
> We would like to ask you to update the firmware to the latest available version so that it can be updated via fwupdmgr, so that no additional windows-vm is required to update the firmware.

Sorry for the (slightly?) off-topic question: What exactly is required hardware-wise to update the dock?

Do you need a Lenovo laptop for that? Does the dock have to be connected to the laptop using Thunderbolt, or will "simple USB" be enough?

Background of my question: I will be using the dock with a Mac soon and may no longer own the Lenovo, so I'm searching for ways to keep the dock's firmware up-to-date even under these circumstances... :-)

Many thanks in advance.

**Comment by hughsie** [comment]
> Do you need a Lenovo laptop for that? Does the dock have to be connected to the laptop using Thunderbolt, or will "simple USB" be enough?

IIUC, it's just a plain USB device, no USB-4 or Lenovo required.

**Comment by ralfbergs** [comment]
> > Do you need a Lenovo laptop for that? Does the dock have to be connected to the laptop using Thunderbolt, or will "simple USB" be enough?
> 
> IIUC, it's just a plain USB device, no USB-4 or Lenovo required.

Thanks a lot for your response. In that case I think I will just try it out with a Ubuntu boot stick... ;-)

**Comment by DerDreschner** [comment]
@mrhpearson Any updates on when we can expect the update to appear in LVFS?

**Comment by mrhpearson** [comment]
Nagging the team...I'm not seeing any updates uploaded yet and they're supposed to be there. Not sure if they hit any problems or not so I'll find out what the latest is

**Comment by vifino** [comment]
Hey, I'd love to see this hit LVFS as well.

**Comment by VeldoraTheDragon** [comment]
@mrhpearson Any updates on this? 
Version 10.13 release will be one year old in four months. 
Companies that run Linux are relying on this update, getting to LVFS.

**Comment by mrhpearson** [comment]
The team had started testing 10.15 on LVFS, but hit issues with that update (I don't have the details of what issues unfortunately). 
Currently plan is to work on getting 10.16 up this month.

Mark

**Comment by RyanBlakeIT** [comment]
There are definitely monitor issues with 1.0.15 for 40B0.  I've used 3 models of Lenovo laptops with the same behavior and at least 3-4 docking stations.  Screen will go blank even though it reports as connected in the nVIDIA control panel (Discrete Graphics flag set on P16 Gen 2).  So you are definitely better off not getting this update yet, worse than 1.0.13.

**Comment by VeldoraTheDragon** [comment]
@RyanBlakeIT 

> There are definitely monitor issues with 1.0.15 for 40B0. I've used 3 models of Lenovo laptops with the same behavior and at least 3-4 docking stations. Screen will go blank even though it reports as connected in the nVIDIA control panel (Discrete Graphics flag set on P16 Gen 2). So you are definitely better off not getting this update yet, worse than 1.0.13.

Could this problem perhaps be related to this issue's following part?
https://gitlab.freedesktop.org/drm/intel/-/issues/8340
> 5\.At this point, when the dock is connected to the laptop and the monitors are connected to the dock, neither monitor displays an image (yes, that is a problem, but it is not my biggest problem, for now.), but the system shows that they are connected (both via the Ubuntu panel and via xrandr).

**Comment by DerDreschner** [comment]
The new firmware 1.0.16 is available in lvfs: https://fwupd.org/lvfs/devices/com.lenovo.ThinkPad.Thunderbolt4.dock.firmware

@RyanBlakeIT: My T15g Gen 2 had the same issues. I installed the current [Intel Management Engine Firmware](https://pcsupport.lenovo.com/us/en/products/laptops-and-netbooks/thinkpad-t-series-laptops/thinkpad-t15g-gen-2-type-20ys-20yt/20ys/downloads/ds551387-intel-management-engine-150-firmware-for-windows-10-version-1809-or-later-thinkpad-p15-gen-2-p15v-gen-2-p17-gen-2-t15g-gen-2-t15p-gen-2?category=Motherboard%20Devices%20%28core%20chipset,%20onboard%20video,%20PCIe%20switches%29), which was released recently and contains a new thunderbolt firmware. If you have a look at the readme, there are plenty of fixes for display output over thunderbolt. And I can confirm that this update together with the 1.0.16 dock firmware made it quite stable.

Let me know if the new thunderbolt and dock firmware made your experience way more usable like it did for me.

**Comment by benblasco** [comment]
@DerDreschner 1.0.16 still causes problems for me with displayport output on Fedora 38. I have two monitors connected via DisplayPort and they don't wake when resuming from suspend. I also had to change some setting in the laptop BIOS to make the NIC work again. The only remedy I have found is to unplug and replug. Fedora's Gnome "Displays" settings shows that the screens still exist, but they don't get any video signal. Do you have any thoughts about downgrading the firmware? I trhink I was ok with 1.0.13.

**Comment by mrhpearson** [comment]
Are those two 4k monitors? Those won't work reliably - though Intel have just done some fixes that will help with that. Not delivered upstream yet.

**Comment by mbollmann** [comment]
@benblasco I have attempted to downgrade from 1.0.16 to 1.0.13 and it completely bricked my docking station.

**Comment by u1735067** [comment]
@mbollmann does https://forums.lenovo.com/t5/Displays-Options-and-Accessories/Issues-on-DP-ports-with-40B0-Thunderbolt-4-dock/m-p/5245873?page=1#6132647 helps?

**Comment by mbollmann** [comment]
> @mbollmann does [forums.lenovo.com/t5/Displays-Options-and-Accessories/Issues-on-DP-ports-with-40B0-Thunderbolt-4-dock/m-p/5245873?page=1#6132647](https://forums.lenovo.com/t5/Displays-Options-and-Accessories/Issues-on-DP-ports-with-40B0-Thunderbolt-4-dock/m-p/5245873?page=1#6132647) helps?

I don't have the dock in my possession anymore, but when I had, it was not recognized by the laptop in any form (didn't even charge the laptop) after my attempted downgrade.

**Comment by DerDreschner** [comment]
> @mbollmann does https://forums.lenovo.com/t5/Displays-Options-and-Accessories/Issues-on-DP-ports-with-40B0-Thunderbolt-4-dock/m-p/5245873?page=1#6132647 helps?

Well, I don't know what I did - but I were once able to unbrick my dock after an attempted downgrade. Nevertheless, I were not able to tell what exactly lead to my success. The commands in that post seem like what I did. So, yeah, I would say that it's possible to downgrade this way. Thanks for the link!

**Comment by benblasco** [comment]
@mrhpearson Thanks for replying. One is a WQHD screen and the other is just a regular FHD, with the third being my laptop screen. I see you're the person working on a potential fix that may land in kernel 6.8, and I just want to say thank you for the work you do!

@u1735067 These downgrade instructions sound scary, but I will consider them!

@mbollmann I am sorry your dock got bricked... that would be infuriating.

**Comment by VeldoraTheDragon** [comment]
@mrhpearson 

![image](https://github.com/fwupd/firmware-lenovo/assets/127216238/2281786e-6f8f-4f88-8ea9-d438c5ab8fbd)
The version 1.0.16 (10.16) has been updated for Windows on December 20th 2023, but the version for Linux on lvfs, same version (10.16), has been updated on August 10th 2023.

What is the difference between those versions?

**Comment by mrhpearson** [comment]
From the release notes looks like these are admin related:

```
VERSION INFORMATION											
											
package       version      rev    date											
--------------------------------------------------------------------------------	
tbt4dkfw1016  1.0.16        05    2023-12-04
tbt4dkfw1016  1.0.16        04    2023-10-23
tbt4dkfw1016  1.0.16        02    2023-08-02
<snip>						
											
  Note: Revision number (Rev.) is for administrative purpose of this README											
        document and is not related to software version. There is no need to											
        upgrade this software when the revision number changes.	
```
There's some details in the release notes, but looks like just changes to the installer so that it runs on newer platforms (and some minor update notification stuff which doesn't apply to Linux)
I think you're safe to ignore this one.

**Comment by c0d3z3r0** [comment]
@mrhpearson I compared the firmware images (splitted images per controller/mcu) and there are definitely changes. Maybe this needs more investigation or some statement from the FW team.

I was not able yet to identify the two differing images, however FW part 09 (line 10) contains a version number 21.15, which matches the `GD MCU FW version`. No idea what that GD MCU is, though. (It's an ARM binary and there are only two in that dock: CY7C DMC and GGC5, but I suspect the main images to be among the `unknown` images).

Left: most recent update from support site (1.0.16 rev 05)
Right: most recent update from lvfs

```
|   1 1b64db238e3096fc2bd59734c1b1cb9091cd16e9  cy7c65219  |   1 1b64db238e3096fc2bd59734c1b1cb9091cd16e9  cy7c65219
|   2 3d06570961ae5b40d5ebd8b23e1bc4ea2111fec9  fl5801     |   2 3d06570961ae5b40d5ebd8b23e1bc4ea2111fec9  fl5801
|   3 45fd2b5ca3c861e7a895e568dc6547a12e6cbd91  jhl8440    |   3 45fd2b5ca3c861e7a895e568dc6547a12e6cbd91  jhl8440
|   4 6be1b2c065213768181b1c97ee8cae1a95493d60  via hub(s) |   4 6be1b2c065213768181b1c97ee8cae1a95493d60  via hub(s)
|   5 71d0277f0cf48ef6770247f54e0a7cabd1ac7b09  i225       |   5 71d0277f0cf48ef6770247f54e0a7cabd1ac7b09  i225
|   6 89a6ef33d6a4850618c6b1d96d98a239503b6339  unknown    |   6 89a6ef33d6a4850618c6b1d96d98a239503b6339  unknown
|   7 9730c181b044a0828ae6011f19ce41a5a36408d0  cx21985    |   7 9730c181b044a0828ae6011f19ce41a5a36408d0  cx21985
|   8 9e1a8857aad75c8444d79285a541b65e14d77075  unknown    |   8 9e1a8857aad75c8444d79285a541b65e14d77075  unknown
>   9 5f9e16719b639679f40823a6bc49bce27f98bca7  "SL" mcu   |   9 6941efa8022001b0bcadd33cf6a5767b428b883d  "SL" mcu                       
>  10 adb96fb4f0785f3395fe935734e6afac0b052a5d  "GD" mcu   |  10 c40cdd8e7b69c783a5a7c0cdd2ae89bb8c93eef7  "GD" mcu                        
```

**Comment by c0d3z3r0** [comment]
Update: I identified two of the unknown images.

SL UB20F64G appears as "Lenovo ... Dock MCU Controller 0123456789ABCDEF" in lsusb.
"GD" mcu could be an equivalent being used on some boards instead of the SL mcu. It identifies as "Lenovo ... Dock MCU Controller 2". Or maybe it's on the "smart" dock (40B1), but the 40B1 package does not list this "GD mcu".
Having a *very* rough look at the firmware images raises the suspicion these MCUs are only (mainly?) used during the update process. That could explain why the fw team says there is no functional difference. Not sure if they are used by fwupd at all. But why would Windows update need another MCU while fwupd can do it without?

Fun fact: the non-smart 40B0 package contains images for the smart version, even though a separate update package exists. Weird o.O

**Comment by soig** [comment]
@mrhpearson BTW could you provide changelog through LVFS too?
That'd be nice too : I was looking for changelog for that 10.16 update…

**Comment by ralfbergs** [comment]
(I hope it's allowed to post this here -- I suspect some people my use the Linux "fwupd" tool just to update the firmware of their dock, but otherwise don't use Linux...)

I wanted to draw your attention to the recently released 1.0.18 update. It completely broke Mac compatibility for me.

For the details, see [my Reddit post](https://www.reddit.com/r/mac/comments/1dwsjd9/warning_lenovo_universal_thunderbolt_4_dock_40b0/).

**Comment by mrhpearson** [comment]
@solg - changelogs are supposed to be in the metadata for releases. I'll drop a note to the docking team to add notes on the releases - it does seem limited. 

@ralfbergs - Sorry - Mac is out of scope for me, I only deal with our Linux supported platforms - but I'll let the docking team know about the issue.
